### PR TITLE
[rayci] add job nodes, prepare for dependency tracking

### PR DESCRIPTION
**Pull request by aslonnie**
separate parsing, filtering and converting; get ready for dependency tracking and lambda invocation

### CODE REVIEW

File: raycicmd/converter.go [hunk 1]
```diff
@@ -71,25 +71,24 @@ func (c *converter) convertStep(step map[string]any) (
...
 	}
 
-	for _, step := range g.Steps {
-		// filter steps by tags
-		if stepTags, ok := step["tags"]; ok {
-			if !filter.hit(toStringList(stepTags)) {
-				continue
-			}
+	for _, step := range n.steps {
+		if !step.include {
```
Comment (can-anyscale) [inline raycicmd/converter.go:86]: nit: include --> isIncluded? includeSomething? to make it a boolean

File: raycicmd/converter.go [hunk 2]
```diff
@@ -102,10 +101,81 @@ func (c *converter) convertGroup(g *pipelineGroup, filter *tagFilter) (
 func (c *converter) convertGroups(gs []*pipelineGroup, filter *tagFilter) (
 	[]*bkPipelineGroup, error,
 ) {
+
+	var groupNodes []*jobNode
+
+	for i, g := range gs {
+		node := &jobNode{
+			id:       fmt.Sprintf("g%d", i),
+			userKey:  g.Key,
```
Comment (can-anyscale) [inline raycicmd/converter.go:110]: just `key`? what the word `user` mean?

File: raycicmd/converter.go [hunk 3]
```diff
@@ -102,10 +101,81 @@ func (c *converter) convertGroup(g *pipelineGroup, filter *tagFilter) (
...
+				tags:    tags,
+			})
+		}
+
+		groupNodes = append(groupNodes, node)
+	}
+
+	// Build namedNodes, and check if we have duplicated user keys.
+	nameNodes := make(map[string]*jobNode)
+	for _, g := range groupNodes {
```
Comment (can-anyscale) [inline raycicmd/converter.go:139]: nit: g --> n (it's a node not a group object?)

File: raycicmd/converter.go [hunk 4]
```diff
@@ -102,10 +101,81 @@ func (c *converter) convertGroup(g *pipelineGroup, filter *tagFilter) (
...
+				srcStep: step,
+				tags:    tags,
+			})
+		}
+
+		groupNodes = append(groupNodes, node)
+	}
+
+	// Build namedNodes, and check if we have duplicated user keys.
+	nameNodes := make(map[string]*jobNode)
```
Comment (can-anyscale) [inline raycicmd/converter.go:138]: this map is used only to check duplicated keys and not used anywhere else? if so just need a set?

File: raycicmd/job_node.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package raycicmd
+
+// jobNode is a node for a job node or a group in the pipeline.
```
Comment (can-anyscale) [inline raycicmd/job_node.go:3]: is a job map to a buildkite build? `node for a job node` is a bit hard to understand

Comment (can-anyscale) [reply raycicmd/job_node.go:3]: `group` -> `group of steps`

File: raycicmd/job_node.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+type jobNode struct {
+	id   string
+	tags []string
+
+	userKey string // user defined key
+
+	// fields used for groups
+	srcGroup *pipelineGroup // set for group nodes
+	steps    []*jobNode
```
Comment (can-anyscale) [inline raycicmd/job_node.go:12]: childNodes? can the number of concepts here be reduced (job, node, step, group)

**Review by can-anyscale — APPROVED** [review]

File: raycicmd/converter.go (hunk 1 above)
Comment (aslonnie) [reply raycicmd/converter.go]: changed to "marked".

File: raycicmd/converter.go (hunk 4 above)
Comment (aslonnie) [reply raycicmd/converter.go]: done.

I will need something to query user defined names down the road.

**Pull request merged**